### PR TITLE
Issue #1144 - Phase 2 : SIMD related concept + logical constraints

### PR DESCRIPTION
**Pull request by jfalcou**

### CODE REVIEW

File: include/eve/concept/simd.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  namespace detail
+  {
+    template<typename T> struct is_wide { static constexpr bool value = false; };
+
+    template<scalar T, typename N>
+    struct is_wide<wide<T,N>> { static constexpr bool value = true; };
+  }
+
+  template<typename T>
+  concept simd = detail::is_wide<T>::value;
```
Comment (DenisYaroshevskiy) [inline include/eve/concept/simd.hpp:30]: Should we just move instance of to main detail? We have it in algo, we needed it for aligned pointer, now here?

Comment (DenisYaroshevskiy) [reply include/eve/concept/simd.hpp:30]: Wait. Is logical<simd> - simd?

File: include/eve/concept/simd.hpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  template<typename T>
+  concept plain_simd = simd<T> && plain_scalar<element_type_t<T>>;
+
+  template<typename T>
+  concept product_simd  = simd<T> && kumi::product_type<element_type_t<T>>;
+
+  template<typename T>
+  concept plain_value = plain_scalar<T> || plain_simd<T>;
```
Comment (DenisYaroshevskiy) [inline include/eve/concept/simd.hpp:39]: Do the other way around please:

```
concept plain_value = plain_scalar<element_type_t<T>>;
concept plain_simd..
```

**Review by DenisYaroshevskiy — COMMENTED** [review]
I am not wild about "simd" as a concept name. 
Why not simd_value?

**Comment by jfalcou** [comment]
> I am not wild about "simd" as a concept name. Why not simd_value?

As per #1144 discussion and decision, `simd_value` will be used for something else.

On e potential solution is to rename scalar -> vectorizable and simd -> vectorized AFTER we changed all the other concepts.
We can keep simd for now as a transition phase.

File: include/eve/concept/simd.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/concept/simd.hpp:30]: Per #1144, logical<wide< is a simd_value but not simd
simd is basically the builtin stuff+product type as we decided in the #1144 discussion

File: include/eve/concept/simd.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/concept/simd.hpp:39]: `plain_value` is either `plain_simd` or `plain_scalar` so I don't get what you mean

File: include/eve/concept/simd.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/concept/simd.hpp:30]: I'll move instance_of into meta.hpp

File: include/eve/concept/simd.hpp (hunk 2 above)
Comment (DenisYaroshevskiy) [reply include/eve/concept/simd.hpp:32]: Hmm. I wanted `plain_value` to be `plain<element_type_t<T>` but this leads to issues.

ok fair enough.

**Comment by DenisYaroshevskiy** [comment]
Discussion called `instance_of_wide` `arithmetic_simd_value`.
ANd it called `scalar_value` `arithmetic_scalar_value `.

We never agreed on `simd` concept that's for sure.

**Review by DenisYaroshevskiy — APPROVED** [review]

**Pull request merged**